### PR TITLE
http3/quiche: fix invalid request when uploading big file

### PR DESCRIPTION
**Pull request by bachue**
current [master](https://github.com/curl/curl/commit/78af8b68cfe4222057100911c50d52f3719b47d9):
```
$ curl --tlsv1.3 --http3 -vk -X POST --data-binary "@100mb-file" https://blog.cloudflare.com/

...
> POST / HTTP/3
> Host: blog.cloudflare.com
> user-agent: curl/7.74.1-DEV
> accept: */*
> content-length: 104857600
> content-type: application/x-www-form-urlencoded
>
* STATE: DO => DO_DONE handle 0x12cd128; line 1940 (connection #0)
* STATE: DO_DONE => PERFORM handle 0x12cd128; line 2061 (connection #0)
* Pass on 65536 body bytes to quiche
* Kill stream: Transfer returned error
* multi_done
* Connection #0 to host blog.cloudflare.com left intact
* Expire cleared (transfer 0x12cd128)
curl: (55) Failed sending data to the peer
```
after applying this patch:
```
 curl --tlsv1.3 --http3 -vk -X POST --data-binary "@100mb-file" https://blog.cloudflare.com/

...
* STATE: DO => DO_DONE handle 0x71f128; line 1940 (connection #0)
* STATE: DO_DONE => PERFORM handle 0x71f128; line 2061 (connection #0)
...
* We are completely uploaded and fine
* HTTP 1.1 or later with persistent connection
< HTTP/3 404
< date: Mon, 14 Dec 2020 10:03:31 GMT
< content-type: text/html; charset=utf-8
< set-cookie: __cfduid=d4cb754cb871a9c5c274c7243acba52a31607940201; expires=Wed, 13-Jan-21 10:03:21 GMT; path=/; domain=.blog.cloudflare.com; HttpOnly; SameSite=Lax; Secure
< x-powered-by: Express
< cache-control: no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0
< vary: Accept-Encoding
< cf-cache-status: DYNAMIC
< cf-request-id: 07024c0fd50000237496382000000001
< expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< server: cloudflare
< cf-ray: 601715f95e5b2374-HKG
< alt-svc: h3-27=":443"; ma=86400, h3-28=":443"; ma=86400, h3-29=":443"; ma=86400
* Curl_readwrite: forcibly told to drain data
...
```

### CODE REVIEW

File: lib/vquic/quiche.c [hunk 1]
```diff
@@ -404,10 +406,16 @@ static CURLcode flush_egress(struct connectdata *conn, int sockfd,
...
 
-    sent = send(sockfd, out, sent, 0);
-    if(sent < 0) {
-      failf(conn->data, "send() returned %zd\n", sent);
-      return CURLE_SEND_ERROR;
+resend:
+    socksent = send(sockfd, out, sent, 0);
+    if(socksent < 0) {
+      if(errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR) {
+        goto resend;
```
Comment (bagder) [inline lib/vquic/quiche.c:413]: this is a busy-loop... shouldn't this rather wait for the socket to become writable?

This also makes the function annoyingly blocking, which we should fix as well but could of course be done separately.

File: lib/vquic/quiche.c [hunk 2]
```diff
@@ -623,11 +634,16 @@ static CURLcode http_request(struct connectdata *conn, const void *mem,
   for(i = 1; i < len; ++i) {
     if(hdbuf[i] == '\n' && hdbuf[i - 1] == '\r') {
       ++nheader;
+      if(i >= 3 && hdbuf[i - 2] == '\n' && hdbuf[i - 3] == '\r') {
+        break;
```
Comment (bagder) [inline lib/vquic/quiche.c:638]: This seems unrelated to this bugfix, right? Why is this change necessary?

File: lib/vquic/quiche.c [hunk 3]
```diff
@@ -808,7 +849,12 @@ static CURLcode http_request(struct connectdata *conn, const void *mem,
   if(stream3_id < 0) {
     H3BUGF(infof(data, "quiche_h3_send_request returned %d\n",
                  stream3_id));
-    result = CURLE_SEND_ERROR;
+    if(stream3_id == QUICHE_H3_ERR_STREAM_BLOCKED) {
+      result = CURLE_AGAIN;
```
Comment (bagder) [inline lib/vquic/quiche.c:853]: I realize this is pretty much the only way we can deal with this now, but getting stream blocked here and just blindly retry it at the next possible moment is going to cause some crazy busy-looping. I would imagine that this rather needs a more clever back-off mechanism or a timer to make the retry hold off a bit.

File: lib/vquic/quiche.c [hunk 4]
```diff
@@ -831,14 +877,27 @@ CURLcode Curl_quic_done_sending(struct connectdata *conn)
...
     struct quicsocket *qs = conn->quic;
-    fprintf(stderr, "!!! Curl_quic_done_sending\n");
     stream->upload_done = TRUE;
     sent = quiche_h3_send_body(qs->h3c, qs->conn, stream->stream3_id,
                                NULL, 0, TRUE);
-    if(sent < 0)
-      return CURLE_SEND_ERROR;
+    if(sent < 0) {
+      if(sent == QUICHE_H3_ERR_DONE) {
+        return CURLE_AGAIN;
```
Comment (bagder) [inline lib/vquic/quiche.c:889]: This quiche error is "documented" (in the header) as `There is no more work to do.`. Why do we need to EAGAIN it?

File: lib/vquic/quiche.c (hunk 2 above)
Comment (bachue) [reply lib/vquic/quiche.c:638]: @bagder  It's related. I just found if I upload small file, the request body would follow the http header, with two `\r\n` splited, just like the http protocol. In this case, if you don't stop when you meet two continuous `\r\n`, and by coincidence there's another `\r\n` in your request body, then you could get the wrong header count.

Here's my gdb print:

```
$ LD_LIBRARY_PATH=$PWD/lib/.libs/ rust-gdb --args ./src/.libs/curl --tlsv1.3 --http3 -vk -X POST --data-binary "@1k-file" https://blog.cloudflare.com/
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.3) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./src/.libs/curl...done.
(gdb) break quiche.c:633
No source file named quiche.c.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (quiche.c:633) pending.
(gdb) run
Starting program: /root/curl/src/.libs/curl --tlsv1.3 --http3 -vk -X POST --data-binary @1k-file https://blog.cloudflare.com/
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Note: Unnecessary use of -X or --request, POST is already inferred.
* STATE: INIT => CONNECT handle 0x64b548; line 1654 (connection #-5000)
* Added connection 0. The cache now contains 1 members
[New Thread 0x7ffff6691700 (LWP 1215)]
* STATE: CONNECT => WAITRESOLVE handle 0x64b548; line 1700 (connection #0)
[Thread 0x7ffff6691700 (LWP 1215) exited]
* family0 == v4, family1 == v6
*   Trying 104.18.26.46:443...
* Sent QUIC client Initial, ALPN: h3-29,h3-28,h3-27
* STATE: WAITRESOLVE => WAITCONNECT handle 0x64b548; line 1782 (connection #0)
* quiche established connection!
* Connected to blog.cloudflare.com (104.18.26.46) port 443 (#0)
* Marked for [keep alive]: HTTP/3 default
* STATE: WAITCONNECT => SENDPROTOCONNECT handle 0x64b548; line 1848 (connection #0)
* STATE: SENDPROTOCONNECT => DO handle 0x64b548; line 1869 (connection #0)

Breakpoint 1, http_request (conn=0x64a5a8, mem=0x6794f8, len=1181, written=0x7fffffffdcc8) at vquic/quiche.c:633
633       nheader = 0;
(gdb) p hdbuf
$1 = 0x6794f8 "POST / HTTP/3\r\nHost: blog.cloudflare.com\r\nUser-Agent: curl/7.74.0-DEV\r\nAccept: */*\r\nContent-Length: 1024\r\nContent-Type: application/x-www-form-urlencoded\r\n\r\nt8\215\017\223t\352\201\314\323-ZQv\226\336\304\332\356ZV&\323n\330e\341\364V\357\372\317X\324q?\234\272\252\243*+\024"...
```

File: lib/vquic/quiche.c (hunk 4 above)
Comment (bachue) [reply lib/vquic/quiche.c:889]: @bagder I do this because I read the source code of quiche:
```rust
        if stream_cap <= overhead {
            return Err(Error::Done);
        }
```
url: https://github.com/cloudflare/quiche/blob/master/src/h3/mod.rs#L1026 

And the document of this method mentioned:

```
    /// When a partial write happens (including when [`Done`] is returned) the
    /// application should retry the operation once the stream is reported as
    /// writable again.
```

When you upload a big file, the quiche send buffer could be exhausted, and the `stream_capacity`  here could be zero, but the `overhead` will never be zero, so this method `send_body()` would returns `Done`. In this situation, curl should retry later, rather than return `CURLE_SEND_ERROR`.

File: lib/vquic/quiche.c (hunk 1 above)
Comment (bachue) [reply lib/vquic/quiche.c:418]: OK, I have done as what you said in my new commit.

File: lib/vquic/quiche.c [hunk 5]
```diff
@@ -410,6 +412,9 @@ static CURLcode flush_egress(struct connectdata *conn, curl_socket_t sockfd,
     socksent = send(sockfd, out, sent, 0);
     if(socksent < 0) {
       if(errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR) {
+        FD_ZERO(&fds_write);
+        FD_SET(sockfd, &fds_write);
+        Curl_select(sockfd + 1, NULL, &fds_write, NULL, 0);
```
Comment (bagder) [inline lib/vquic/quiche.c:417]: This should not select() by itself since this makes a blocking behavior that excludes all other concurrent transfers while this waits there. We have a system to return EAGAIN and get back to operation when the socket becomes writable again and we should rather make an effort that h3 code can use this as well.

File: lib/vquic/quiche.c (hunk 4 above)
Comment (bagder) [reply lib/vquic/quiche.c:894]: I think that fairly unusual mandate could warrant a comment next to the code I asked about so that the next fellow developer reading that part of the code can get an explanation for it as well!

File: lib/vquic/quiche.c (hunk 5 above)
Comment (bachue) [reply lib/vquic/quiche.c]: OK, now I return CURLE_AGAIN when the sockfd here is blocking.

**Comment by bagder** [comment]
> "This branch cannot be rebased due to conflicts"

Can you please rebase this on top of master and force-push?

**Comment by bachue** [comment]
> > "This branch cannot be rebased due to conflicts"
> 
> Can you please rebase this on top of master and force-push?

Done

File: lib/vquic/quiche.c [hunk 6]
```diff
@@ -246,6 +251,10 @@ CURLcode Curl_quic_connect(struct connectdata *conn, curl_socket_t sockfd,
 
   result = flush_egress(conn, sockfd, qs);
   if(result) {
+    if(result == CURLE_AGAIN) {
+      /* Curl_quic_connect() should never returns CURLE_AGAIN */
```
Comment (bagder) [inline lib/vquic/quiche.c:255]: If it shouldn't, then how come it can? And why couldn't it? You have the same comment and code in `Curl_quic_is_connected()` too. The comment and logic is weird and I think it deserves more explanation!

File: lib/vquic/quiche.c (hunk 5 above)
Comment (bagder) [inline lib/vquic/quiche.c]: This can't be right. If if can't send more now, it should do the non-blocking behavior and return EAGAIN all the way back and get called again later.

**Review by bagder — COMMENTED** [review]
Can you please squash the commits too? It's very hard to review code to the same code base done in multiple commits like this.

File: lib/vquic/quiche.c (hunk 6 above)
Comment (bachue) [reply lib/vquic/quiche.c]: Ok, I have commented.

**Comment by bachue** [comment]
> Can you please squash the commits too? It's very hard to review code to the same code base done in multiple commits like this.

Done

File: lib/vquic/quiche.c [hunk 7]
```diff
@@ -773,28 +870,44 @@ static CURLcode http_request(struct Curl_easy *data, const void *mem,
...
   switch(data->state.httpreq) {
   case HTTPREQ_POST:
   case HTTPREQ_POST_FORM:
   case HTTPREQ_POST_MIME:
   case HTTPREQ_PUT:
-    if(data->state.infilesize != -1)
-      stream->upload_left = data->state.infilesize;
-    else
+    if(data->state.infilesize != -1) {
+      stream->upload_left = len - ((char *)hdbuf - (char *)mem);
```
Comment (bagder) [inline lib/vquic/quiche.c:881]: This changes how `stream->upload_left` is used to now be how much memory there is in the current buffer to send. Is that on purpose?

Looking at this closer, we don't even use `->upload_left` elsewhere in quiche.c. Should we just use a local variable there instead?

**Review by bagder — COMMENTED** [review]
Thanks for your work on this and sorry for the time it's taken me to come back. Here's a bunch of new comments on the current work.

**Comment by bagder** [comment]
~~~
./vquic/quiche.c:606:5: warning: if with space (SPACEBEFOREPAREN)
   if (stream->upload_done) {
     ^
./vquic/quiche.c:608:3: warning: else after closing brace on same line (BRACEELSE)
   } else {
   ^
./vquic/quiche.c:662:5: warning: else after closing brace on same line (BRACEELSE)
     } else {
     ^
./vquic/quiche.c:663:97: warning: Longer than 79 columns (LONGLINE)

**Comment by bachue** [comment]
@bagder sorry I'm still working on a bug when upload multiple big files via quiche & curl, may commit several times. I will notify you after I'm ready

**Comment by bagder** [comment]
Closing due to inactivity. When you come back around to working on this, please rebase and create a new PR.
